multiversx_sdk_wallet/user_wallet.py

```diff
@@ -1,18 +1,78 @@
 import json
+from enum import Enum
 from pathlib import Path
-from typing import Any, Dict, Union
+from typing import Any, Dict, Optional, Union
 
-from multiversx_sdk_wallet import keyfile
-from multiversx_sdk_wallet.interfaces import IUserWalletRandomness
-from multiversx_sdk_wallet.user_keys import UserSecretKey
+from multiversx_sdk_wallet.crypto import (EncryptedData, Randomness, decryptor,
+                                          encryptor)
+from multiversx_sdk_wallet.interfaces import IRandomness
+from multiversx_sdk_wallet.mnemonic import Mnemonic
+from multiversx_sdk_wallet.user_keys import UserPublicKey, UserSecretKey
+
+
+class UserWalletKind(str, Enum):
+    SECRET_KEY = "secretKey"
+    MNEMONIC = "mnemonic"
 
 
 class UserWallet:
-    def __init__(self, secret_key: UserSecretKey, password: str, randomness: Union[IUserWalletRandomness, None] = None) -> None:
-        self.secret_key = secret_key
-        self.public_key = secret_key.generate_public_key()
-        self.password = password
-        self.randomness = randomness
+    def __init__(self, kind: str, encrypted_data: EncryptedData, public_key_when_kind_is_secret_key: Optional[UserPublicKey] = None) -> None:
+        """
+        Do not use this constructor directly. Use the static methods from_secret_key() and from_mnemonic() instead.
+        """
+        self.kind = kind
+        self.encrypted_data = encrypted_data
+        self.public_key_when_kind_is_secret_key = public_key_when_kind_is_secret_key
+
+    @classmethod
+    def from_secret_key(cls, secret_key: UserSecretKey, password: str, randomness: Union[IRandomness, None] = None) -> 'UserWallet':
+        randomness = randomness or Randomness()
+
+        public_key = secret_key.generate_public_key()
+        data = secret_key.buffer + public_key.buffer
+        encrypted_data = encryptor.encrypt(data, password, randomness)
+
+        return cls(
+            kind=UserWalletKind.SECRET_KEY,
+            encrypted_data=encrypted_data,
+            public_key_when_kind_is_secret_key=public_key
+        )
+
+    @classmethod
+    def from_mnemonic(cls, mnemonic: str, password: str, randomness: Union[IRandomness, None] = None) -> 'UserWallet':
+        randomness = randomness or Randomness()
+
+        Mnemonic.assert_text_is_valid(mnemonic)
+        data = mnemonic.encode()
+        encrypted_data = encryptor.encrypt(data, password, randomness)
+
+        return cls(
+            kind=UserWalletKind.MNEMONIC,
+            encrypted_data=encrypted_data
+        )
+
+    @classmethod
+    def decrypt_secret_key(cls, keyfile_object: Dict[str, Any], password: str) -> UserSecretKey:
+        # Here, we do not check the "kind" field. Older keystore files (holding only secret keys) do not have this field.
+
+        encrypted_data = EncryptedData.from_keyfile_object(keyfile_object)
+        buffer = decryptor.decrypt(encrypted_data, password)
+
+        while len(buffer) < 32:
+            buffer = b'\x00' + buffer
```

Comment:
Maybe [bytes.rjust()](https://docs.python.org/3.8/library/stdtypes.html#bytes.rjust) could be used instead.

Reply:
Fixed.